File: lib/market_data/types/symbol.ex
defmodule MarketData.Types.Symbol do
  @moduledoc """
  Symbol type
  """
  @type t :: %__MODULE__{
          symbol: String.t(),
          baseAsset: String.t(),
          quoteAsset: String.t(),
          baseAssetPrecision: integer(),
          quoteAssetPrecision: integer(),
          baseAssetIncrement: Decimal.t(),
          quoteAssetIncrement: Decimal.t(),
          minNotional: Decimal.t()
        }

  @enforce_keys [
    :symbol,
    :baseAsset,
    :quoteAsset,
    :baseAssetPrecision,
    :quoteAssetPrecision,
    :baseAssetIncrement,
    :quoteAssetIncrement,
    :minNotional
  ]
  defstruct [
    :symbol,
    :baseAsset,
    :quoteAsset,
    :baseAssetPrecision,
    :quoteAssetPrecision,
    :baseAssetIncrement,
    :quoteAssetIncrement,
    :minNotional
  ]
end

File: src/Symbol.cpp
#include "Symbol.h"

Symbol::Symbol(const std::string symbolStr, const std::string baseAsset,
               const std::string quoteAsset, const int baseAssetPrecision,
               const int quoteAssetPrecision, const PreciseNumber minNotional,
               const PreciseNumber baseAssetIncrement,
               const PreciseNumber quoteAssetIncrement)
    : symbol(symbolStr), baseAsset(baseAsset), quoteAsset(quoteAsset),
      minNotional(minNotional), baseAssetIncrement(baseAssetIncrement),
      quoteAssetIncrement(quoteAssetIncrement),
      baseAssetPrecision(baseAssetPrecision),
      quoteAssetPrecision(quoteAssetPrecision) {};